src/lib/RegistrationLib.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+import { ResetPeriod } from "../types/ResetPeriod.sol";
+
+import { EfficiencyLib } from "./EfficiencyLib.sol";
+import { IdLib } from "./IdLib.sol";
+
+/**
+ * @title RegistrationLib
+ * @notice Library contract implementing logic for registering compact claim hashes
+ * and typehashes and querying for whether given claim hashes and typehashes have
+ * been registered.
+ */
+library RegistrationLib {
+    using RegistrationLib for address;
+    using EfficiencyLib for uint256;
+    using IdLib for ResetPeriod;
+
+    // keccak256(bytes("CompactRegistered(address,bytes32,bytes32,uint256)")).
+    uint256 private constant _COMPACT_REGISTERED_SIGNATURE = 0xf78a2f33ff80ef4391f7449c748dc2d577a62cd645108f4f4069f4a7e0635b6a;
+
+    // Storage scope for active registrations:
+    // slot: keccak256(_ACTIVE_REGISTRATIONS_SCOPE ++ sponsor ++ claimHash ++ typehash) => expires.
+    uint256 private constant _ACTIVE_REGISTRATIONS_SCOPE = 0x68a30dd0;
+
+    /**
+     * @notice Internal function for registering a claim hash with a specific duration. The
+     * claim hash and its associated typehash will remain valid until the specified duration
+     * has elapsed. Reverts if the duration would result in an expiration earlier than an
+     * existing registration or if it exceeds 30 days.
+     * @param sponsor   The account registering the claim hash.
+     * @param claimHash A bytes32 hash derived from the details of the compact.
+     * @param typehash  The EIP-712 typehash associated with the claim hash.
+     * @param duration  The duration in seconds for which the registration remains valid.
+     */
+    function registerCompactWithSpecificDuration(address sponsor, bytes32 claimHash, bytes32 typehash, uint256 duration) internal {
+        assembly ("memory-safe") {
+            // Retrieve the current free memory pointer.
+            let m := mload(0x40)
+
+            // Pack data for deriving active registration storage slot.
+            mstore(add(m, 0x14), sponsor)
+            mstore(m, _ACTIVE_REGISTRATIONS_SCOPE)
+            mstore(add(m, 0x34), claimHash)
+            mstore(add(m, 0x54), typehash)
+
+            // Derive and load active registration storage slot to get current expiration.
+            let cutoffSlot := keccak256(add(m, 0x1c), 0x58)
+
+            // Compute new expiration based on current timestamp and supplied duration.
+            let expires := add(timestamp(), duration)
+
+            // Ensure new expiration does not exceed current and duration does not exceed 30 days.
```

Comment:
Should probably be:
            // Ensure new expiration `is not earlier than current` and duration does not exceed 30 days.

Reply:
good catch!